src/lsm/composite_key.zig

```diff
@@ -35,29 +33,31 @@ pub fn CompositeKeyType(comptime Field: type) type {
         field: Field align(field_bitsize_alignment),
         /// The most significant bit must be unset as it is used to indicate a tombstone.
         timestamp: u64,
-        padding: Pad = pad,
+        padding: Pad = 0,
 
         comptime {
-            assert(@sizeOf(Self) == @sizeOf(Field) * 2);
-            assert(@sizeOf(Self) == @sizeOf(Key));
-            assert(@alignOf(Self) >= @alignOf(Field));
-            assert(@alignOf(Self) == field_bitsize_alignment);
-            assert(stdx.no_padding(Self));
+            assert(@sizeOf(CompositeKey) == @sizeOf(Field) * 2);
+            assert(@sizeOf(CompositeKey) == @sizeOf(Key));
+            assert(@alignOf(CompositeKey) >= @alignOf(Field));
+            assert(@alignOf(CompositeKey) == field_bitsize_alignment);
+            assert(stdx.no_padding(CompositeKey));
         }
 
-        pub inline fn key_from_value(value: *const Self) Key {
+        pub inline fn key_from_value(value: *const CompositeKey) Key {
+            assert(value.padding == 0);
```

Comment:
My gut feeling is to guard these asserts with `if constants.verify`: they are going to be in the innermost loop possible, and feels like they _could_ affect codegen adversely. Eg, if you have a loop, the compiler might be able to vectorize it. But if there's a possibility of an early exit from the loop, vectorization might be incorrect (as the code won't actually get to the end of the loop), so compiler might be forced to pick a scalar implementation.

But maybe this assert doesn't make optimization situation worse? We are already running with overflow checks, so an extra `noreturn` possibility _shouldn't_ make the code slower? Hard to tell though! 

I guess my decision tree is this:

* if we have a set of benchmarks that can detect realatively fine-grained performance effects, then we should run those benches. 
* if not, I'd lean towards guarding by `if constants.verify`. 
* Perhaps it would also be interesting to look at the actual optimized wasm of say, binary search, to understand how different it will be with tombstones?

Reply:
>I didn't get it. What relation do tombstones have with the assert on CompositeKey.padding == 0?

Sorry, I think it was a data race between me writing this comment, and me concurently continuing reading the code, where we _do_ have a pre-existing assert in `tombstone_from_key`.

What I meant to say is that looking at the asm forthe inner loop of the binary search over composite keys with and without the assert could be interesting